deps/verifier/src/nvidia/cert_chain.rs

```diff
@@ -0,0 +1,164 @@
+/// issuer=serialNumber = 41431D480FE5E87274, C = US, O = NVIDIA Corporation, CN = GH100 A01 GSP BROM
+///
+/// subject=serialNumber = 41431D480FE5E87274, C = US, O = NVIDIA Corporation, CN = GH100 A01 GSP BROM
+/// issuer=CN = NVIDIA GH100 Provisioner ICA 1, O = NVIDIA Corporation, C = US
+///
+/// subject=CN = NVIDIA GH100 Provisioner ICA 1, O = NVIDIA Corporation, C = US
+/// issuer=CN = NVIDIA GH100 Identity, O = NVIDIA Corporation
+///
+/// subject=CN = NVIDIA GH100 Identity, O = NVIDIA Corporation
+/// issuer=CN = NVIDIA Device Identity CA, O = NVIDIA
+///
+/// subject=CN = NVIDIA Device Identity CA, O = NVIDIA
+/// issuer=CN = NVIDIA Device Identity CA, O = NVIDIA
+pub struct NvidiaCertificateChain {
+    certs: Vec<X509>,
+}
+
+impl NvidiaCertificateChain {
+    pub fn decode(bytes: &[u8]) -> Result<Self> {
+            let certs = X509::stack_from_pem(bytes)?;
+            Ok(Self { certs })
+    }
+
+    pub fn check_root_ca_is_trusted(&self) -> Result<()> {
+        let root_ca = self.get_root_certificate()?;
+        let digest = certificate_fingerprint(root_ca)?;
+
+        if digest != NVIDIA_ROOT_CA.digest {
+            bail!("Root CA in the certificate chain is not trusted");
+        }
+        Ok(())
+    }
+
+    pub fn get_root_certificate(&self) -> Result<&X509> {
+        self.certs
+            .last()
+            .ok_or(anyhow!("Root CA certificate not found in the NVIDIA certificate chain"))
+    }
+
+    pub fn get_leaf_certificate(&self) -> Result<&X509> {
+        self.certs
+            .get(0)
+            .ok_or(anyhow!("Leaf certificate not found in the NVIDIA certificate chain"))
+    }
+
+    /// Use OpenSSL to cryptographically verify the NVIDIA certificate chain.
+    /// 
+    /// Returns the signing certificate if its chain of trust can be verified all the
+    /// way up to the trusted NVIDIA root CA.
+    pub fn verify(&self, expected_fwid: &str) -> Result<&X509> {
+        if self.certs.len() < 2 {
+            bail!("Certificate chain must be >= 2");
+        }
+
+        self.check_root_ca_is_trusted()?;
+
+        // Signing certificate to be verified its chain of trust
+        let leaf_cert = self.get_leaf_certificate()?;
+
+        // Check if leaf certificate FwId matches the expected one
+        let cert_fwid = get_fwid_from_cert(leaf_cert)?;
+        if cert_fwid != expected_fwid {
+            bail!("Fwid mismatch: certificate {}, evidence {}", cert_fwid, expected_fwid);
+        }
+
+        // Trusted root certificate
+        let trusted_certs = {
+            let mut builder = X509StoreBuilder::new().unwrap();
+            builder.add_cert(NVIDIA_ROOT_CA.cert.clone())?;
+            builder.build()
+        };
+
+        // Untrusted certificate chain
+        // OpenSSL 1.1.0+ considers the root certificate to not be part of the chain, while 1.0.2 and LibreSSL do
```

Comment:
how this translates to SSL library versions people *should be using*? would it be better to highlight the part that is specific to LibSSL (if we want to support it)

Reply:
I think we only care about openssl, so I changed the comment to `OpenSSL 1.1.0+ considers the root certificate to not be part of the chain`

For this code to work properly (and possibly the token code https://github.com/confidential-containers/trustee/blob/main/kbs/src/token/jwk.rs#L167), we may need to ensure that we are linking to the right openssl version, for example, we could add somethink like https://docs.rs/openssl/latest/openssl/#feature-detection to the build.rs. That is probably out of scope for this PR ;-)